Dockerfile

```diff
@@ -1,43 +1,86 @@
-FROM tiangolo/uvicorn-gunicorn-fastapi:python3.10-slim
+# Use a multi-stage build to first get uv
+FROM ghcr.io/astral-sh/uv:0.5.10 AS uv
 
-# Comment to trigger an image rebuild
+FROM ubuntu:noble AS build
 
-# Optional build argument for different environments
-ARG ENV
+RUN apt-get update -qy && \
+    apt-get install -qyy \
+        -o APT::Install-Recommends=false \
+        -o APT::Install-Suggests=false \
+        ca-certificates \
+        clang \
+        gcc \
+        git \
+        libgdal-dev \
+        libpq-dev \
+        make
 
-RUN apt-get update -y \
-    && apt-get install --no-install-recommends -y gcc g++ libc-dev \
-        postgresql-client libpq-dev make git jq libgdal-dev \
-    && apt-get clean \
-    && rm -rf /var/lib/apt/lists/*
+# We need to set this environment variable so that uv knows where
+# the virtual environment is to install packages
+ENV UV_LINK_MODE=copy \
+    UV_COMPILE_BYTECODE=1 \
+    UV_PROJECT_ENVIRONMENT=/app/.venv \
+    VIRTUAL_ENV=/app/.venv
 
-RUN pip install --upgrade pip && pip install pipenv==2024.0.1
-#TODO move to pipfile when operational
-RUN pip install newrelic
+# Create a virtual environment with uv inside the container
+RUN --mount=type=cache,target=/app/.cache \
+    --mount=from=uv,source=/uv,target=./uv \
+    /uv python install 3.10 && \
+    /uv venv $VIRTUAL_ENV
 
-# Install python dependencies
-# Install everything for dev and test otherwise just core dependencies
-COPY Pipfile Pipfile
-COPY Pipfile.lock Pipfile.lock
+# Make sure that the virtual environment is in the PATH so
+# we can use the binaries of packages that we install such as pip
+# without needing to activate the virtual environment explicitly
+ENV PATH=$VIRTUAL_ENV/bin:/usr/local/bin:$PATH
 
-RUN if [ "$ENV" = "dev" ] || [ "$ENV" = "test" ]; then \
-        echo "Install all dependencies" \
-        && pipenv install --system --deploy --ignore-pipfile --dev; \
-    else \
-        echo "Install production dependencies only" \
-        && pipenv install --system --deploy; \
-    fi
+RUN --mount=type=cache,target=/app/.cache \
+    --mount=from=uv,source=/uv,target=./uv \
+    /uv pip install setuptools wheel
 
-COPY ./app /app/app
+# Copy pyproject.toml and uv.lock to a temporary directory
+COPY pyproject.toml /_lock/
+COPY uv.lock /_lock/
 
-COPY alembic.ini /app/alembic.ini
+# Install the packages with uv using --mount=type=cache to cache the
+# downloaded packages
+RUN --mount=type=cache,target=/app/.cache \
+    --mount=from=uv,source=/uv,target=./uv \
+    cd /_lock && \
+    /uv sync --locked --no-install-project
+
+# Start the runtime stage
+FROM ubuntu:noble
+SHELL ["sh", "-exc"]
+
+ENV PATH=$VIRTUAL_ENV/bin:/usr/local/bin:$PATH
```

Comment:
Just checking - are all the environment variables (especially VIRTUAL_ENV) that were set during the build stage, also automatically set in the same way for the runtime stage?  I guess so, since otherwise tests wouldn't work, but it seems surprising, since you don't really want some of the build env variables set in the runtime docker.